### PR TITLE
Extension Methods for Sockets

### PR DESCRIPTION
**Pull request by josguil**
Separating legacy methods of Socket as extension methods.

### CODE REVIEW

**Comment by josguil** [comment]
@CIPop  @SidharthNabar @davidsh @pgavlin PTAL .

File: src/System.Net.Sockets.Legacy/ref/System.Net.Sockets.Legacy.csproj [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    <OutputType>Library</OutputType>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="System.Net.Sockets.Legacy.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="project.json" />
+  </ItemGroup>
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.targets))\dir.targets" />
+</Project>
```
Comment (CIPop) [inline src/System.Net.Sockets.Legacy/ref/System.Net.Sockets.Legacy.csproj:15]: NIT: Missing new line at EOF.

File: src/System.Net.Sockets/src/System.Net.Sockets.csproj [hunk 2]
```diff
@@ -230,5 +231,4 @@
     <None Include="project.json" />
   </ItemGroup>
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.targets))\dir.targets" />  
-</Project>
-
+</Project>
```
Comment (CIPop) [inline src/System.Net.Sockets/src/System.Net.Sockets.csproj:234]: NIT: missing new line at EOF.

**Comment by CIPop** [comment]
Thanks Memo! 

Most comments are regarding CodingStyle and the SLN.
There is one thing that the PR seems to be missing: removing certain APIs that are sync-only and some that may cause big issues if left in for NetNative (such as certain `TcpClient` ctors.)

**Comment by SidharthNabar** [comment]
@KrzysztofCwalina @terrajobst @ericstj @weshaggard  - Please take a look at this PR. This is an effort to separate out the APM-style methods in Sockets contract to a different, legacy contract as extension methods. The forward-looking version of Sockets will not have these methods.

We would like your feedback on this from an API review standpoint.

File: src/System.Net.Sockets.Legacy/ref/System.Net.Sockets.Legacy.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+// ------------------------------------------------------------------------------
+// Changes to this file must follow the http://aka.ms/api-review process.
+// ------------------------------------------------------------------------------
+
+
+namespace System.Net.Sockets
+{
+   
+    public static class Legacy
```
Comment (weshaggard) [inline src/System.Net.Sockets.Legacy/ref/System.Net.Sockets.Legacy.cs:11]: We should have a better name for this class. The general pattern is <something>Extensions for static classes with extensions.

File: src/System.Net.Sockets.Legacy/ref/System.Net.Sockets.Legacy.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        public static System.IAsyncResult BeginAccept(this System.Net.Sockets.Socket thisSocket, System.AsyncCallback callback, object state) { return default(System.IAsyncResult); }
```
Comment (weshaggard) [inline src/System.Net.Sockets.Legacy/ref/System.Net.Sockets.Legacy.cs:13]: You can use our GenAPI tool to generate these from the library to help ensure they exactly match your implementation.

**Comment by weshaggard** [comment]
A few comments:
1) We generally avoid using the name Legacy as it doesn't help describe what it is. From the looks this appears to be mostly the old Async stuff so I would probably name it something related to that.
2) Did we ship the System.Net.Sockets 4.0.10 contract anywhere yet? If so we cannot just remove these APIs.
3) How do you plan to implement this new contract? Currently they are implemeneted directly in System.Net.Sockets and you made all the methods in System.Net.Sockets internal, I want to make sure your plan isn't to use InternalsVisibleTo in another library. Also we generally try to not ship multiple contracts in the same implementation when the implementation assembly identity matches that of on of the public contracts. That is generally where we have things like System.Private.Network where we have a private implementation.

**Comment by KrzysztofCwalina** [comment]
We call these APIs "classic async pattern". And so maybe the type should be called ClassicAsyncExtensions?

File: src/System.Net.Sockets.Legacy/ref/System.Net.Sockets.Legacy.cs (hunk 3 above)
Comment (CIPop) [inline src/System.Net.Sockets.Legacy/ref/System.Net.Sockets.Legacy.cs:11]: +1 - I missed this - Originally we were thinking of `SocketLegacyExtensions`.

**Comment by SidharthNabar** [comment]
Thanks Wes and Krzysztof! We will work on a revision of this PR to address # 1 in Wes's list. For the other questions - we haven't shipped 4.0.10.0 yet, except in beta releases of ASP.NET 5where it was still marked experimental. Lastly, for the implementation, in the short term, we were planning to have a façade that type forwards to the implementation of System.Net.Sockets. In the long term, we will extract the source code to a separate DLL.

Thanks,
Sid

**Comment by CIPop** [comment]
> We call these APIs "classic async pattern". 

@KrzysztofCwalina I could find only one reference to ["classic async pattern"](https://www.bing.com/search?q=classic%20async%20pattern&form=EDGENT&qs=PF&cvid=8b29a57b75124abcb93b99846c3185d6&pq=classic%20async%20pattern). MSDN specifies that there are 3 async patterns supported in .Net: EAP, APM and TPL. I've found "Traditional" in [one of the documents](https://msdn.microsoft.com/en-us/library/dd997423%28v=vs.110%29.aspx). Given the changes in CoreFX to _only_ allow TPL for `Stream` and the UWP APIs I doubt that we should be using words like "classic" and "traditional" which can imply these are still supported and should be used. 

Would `System.Net.Sockets.DeprecatedAPM` or `System.Net.Sockets.LegacyAPM`  be better names for the contract?

**Comment by KrzysztofCwalina** [comment]
I thought we don't want to use words "legacy", "deprecated", etc. per statement by @weshaggard above.

**Comment by CIPop** [comment]
> 1) We generally avoid using the name Legacy _as it doesn't help describe what it is_. From the looks this appears to be mostly the old Async stuff so _I would probably name it something related to that_.

@KrzysztofCwalina I think I read it differently :). I thought we don't want to use just "legacy" because it doesn't describes what it is. The new proposed names include the reason (APM APIs).

**Comment by KrzysztofCwalina** [comment]
@weshaggard, could you clarify?

**Comment by weshaggard** [comment]
The guidance is to avoid using things like Legacy in API names. We should try to make the name represent the technology as opposed to names like "New"/"Old" or "Modern"/"Legacy", because those are all relative to a point in time and don't help in identifying the actual technology.

**Comment by davidsh** [comment]
+1 to this:

> The guidance is to avoid using things like Legacy in API names. We should try to make the name represent the technology as opposed to names like "New"/"Old" or "Modern"/"Legacy", because those are all relative to a point in time and don't help in identifying the actual technology. 

The name should be factual and not subjective.  Names like "Modern" and "Legacy" are subjective and judgemental based on a point in time whether they are "good" or "bad" APIs.  So, I agree that we should just be factual and use names to represent that the APIs are "APM async style".

**Comment by CIPop** [comment]
A new PR has been submitted for dotnet master: #3655.
Closing this one.

**Pull request closed without merging**